lib/error_report_notify.c

```diff
@@ -0,0 +1,157 @@
+ *
+ */
+
+#include <linux/fs.h>
+#include <linux/string.h>
+#include <linux/tracepoint.h>
+#include <linux/workqueue.h>
+#include <trace/events/error_report.h>
+#include <trace/events/printk.h>
+
+static struct kobject *error_report_kobj;
+static atomic_t num_error_reports;
+
+static char current_report[PAGE_SIZE], last_report[PAGE_SIZE];
+static size_t current_report_pos, last_report_size;
+static atomic_t current_pid = ATOMIC_INIT(-1);
+/* Protects last_report and last_report_size. */
+static DEFINE_SPINLOCK(last_report_lock);
+
+static struct delayed_work reporting_done;
+
+static void error_report_notify(struct work_struct *work)
+{
+	sysfs_notify(error_report_kobj, NULL, "last_report");
+	sysfs_notify(error_report_kobj, NULL, "report_count");
+}
+static DECLARE_DELAYED_WORK(reporting_done, error_report_notify);
+
+/*
+ * Trace hook for the error_report_start event. If two reports overlap (the
+ * second one report is started before the first one is finished, the first
+ * report is discarded.
+ */
+static void probe_report_start(void *ignore, const char *tool_name,
+			       unsigned long id)
+{
+	atomic_set_release(&current_pid, current->pid);
+}
+
+/*
+ * Trace hook for the error_report_end event. If a stale event from the
+ * mismatching error_report_start is received, it is ignored.
+ */
+static void probe_report_end(void *ignore, const char *tool_name,
+			     unsigned long id)
+{
+	unsigned long flags;
+	pid_t pid = atomic_read_acquire(&current_pid);
+
+	if (!current || pid != current->pid)
+		return;
+	spin_lock_irqsave(&last_report_lock, flags);
+	if (current_report_pos)
+		memcpy(last_report, current_report, current_report_pos);
+	last_report_size = current_report_pos;
+	if (last_report_size == sizeof(last_report))
+		last_report[last_report_size - 1] = 0;
+	else
+		last_report[last_report_size] = 0;
+	spin_unlock_irqrestore(&last_report_lock, flags);
+	atomic_inc(&num_error_reports);
+	schedule_delayed_work(&reporting_done, 0);
+	atomic_set_release(&current_pid, -1);
+}
+
+/*
+ * Trace hook for the console event. If a line comes from a task/CPU that did
+ * not send the error_report_start event, that line is ignored.
+ */
+static void probe_console(void *ignore, const char *buf, size_t len)
+{
+	size_t to_copy;
+	pid_t pid = atomic_read(&current_pid);
+
+	if (!current || pid != current->pid)
+		return;
+
+	to_copy = min(len, sizeof(current_report) - current_report_pos);
+	memcpy(current_report + current_report_pos, buf, to_copy);
+	current_report_pos += to_copy;
+}
+
+static void register_tracepoints(void)
+{
+	register_trace_console(probe_console, NULL);
+	register_trace_error_report_start(probe_report_start, NULL);
+	register_trace_error_report_end(probe_report_end, NULL);
+}
+
+static ssize_t last_report_show(struct kobject *kobj,
+				struct kobj_attribute *attr, char *buf)
+{
+	ssize_t ret;
+	unsigned long flags;
+
+	spin_lock_irqsave(&last_report_lock, flags);
+	ret = scnprintf(buf, PAGE_SIZE, last_report);
```

Comment:
Why does it have to use scnprintf? Can't this just memcpy?

Reply:
We need to return the number of characters written, and scnprintf will do that for us :)